libbeat/outputs/logstash/async.go

```diff
@@ -166,32 +179,37 @@ func (r *msgRef) callback(seq uint32, err error) {
 func (r *msgRef) done(n uint32) {
 	ackedEvents.Add(int64(n))
 	outputs.AckedEvents.Add(int64(n))
-	r.batch = r.batch[n:]
+	r.slice = r.slice[n:]
 	r.win.tryGrowWindow(r.batchSize)
 	r.dec()
 }
 
 func (r *msgRef) fail(n uint32, err error) {
 	ackedEvents.Add(int64(n))
 	outputs.AckedEvents.Add(int64(n))
-	r.err = err
-	r.batch = r.batch[n:]
+
+	if r.err == nil {
```

Comment:
code is correct. It's saying: set `r.err = err` if and only if `r.err` is not set yet. We report the first error only, as upcoming errors are most likely report about connection being closed, but root-cause is first error.